### PR TITLE
nixos/usbguard: rework

### PR DESCRIPTION
**Pull request by tnias**
Override the read only restriction for the preStart script, otherwise
directories in /var/{lib,log} can not be created.

###### Motivation for this change

Fresh usbguard setup with default settings fails to start.

###### Things done

- [x] fixed issue
- [x] tested on my machine

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by tnias** [comment]
cc @ivan @dtzWill @flokli

**Comment by flokli** [comment]
Instead of doing this in the preStart script, can we use the `StateDirectory=` directory to set up these folders, similar to how it's done since https://github.com/NixOS/nixpkgs/pull/93001/files#diff-c485e98c40325e0a3863c8d2de2901a0R167 for gitolite?

If that folder deviates from the default, we don't really know how to set up the file system structure towards this anyways (mount points in between, etc.) - so providing a simple state directory initialization via StateDirectory for the default case, and leaving more complex things up to the user seems preferrable.

Also, cc @aanderse as the author of that original PR.

**Comment by flokli** [comment]
Ah, and `LogsDIrectory=` for the `auditFilePath` - if we can't redirect it to just be put to the systemd journal, which would even be preferrable.

**Comment by tnias** [comment]
I am in favour of using StateDirectory and logging to journal. In that case the some of the options for custom directory paths should be dropped. I will create a patch.

**Comment by flokli** [comment]
Thanks!

**Comment by tnias** [comment]
I could not figure out how to get the audit logs some other way than to write them to file. The normal usbguard logs are already written to the journal. If the audit logs should still land in the linux audit system, I would need help with that, as I did not yet figure out how to get it working (especially on NixOS something with audit/auditd seems boken or unintuitive).

As for dropping the options: Is hardcoding the defaults and removing the options a proper way to deprecate them? Where do I put deprecation notices?

**Comment by aanderse** [comment]
@tnias if you're not hard coding paths how about something like [this](https://github.com/NixOS/nixpkgs/compare/master...aanderse:usbguard)?
@flokli is this along the lines of what you were thinking?

**Comment by flokli** [comment]
I'd even be find with dropping the options, as now proposed in this PR tbh.

**Comment by aanderse** [comment]
@flokli I would also support hard coding the paths so we could drop the options, if everyone else agrees :+1:

**Comment by tnias** [comment]
If I set `AuditFilePath=/var/log/usbguard/usbguard-audit.log` an audit log gets written to the specified file _and_ the journal. If I do not specify a file path, there are no more audit logs in the journal.
With `AuditFIlePath=/dev/null` the journal audit logs are still there and we can drop the file in /var/log.

Also some of the options were dropped and text reworked.

File: nixos/modules/services/security/usbguard.nix [hunk 1]
```diff
@@ -1,37 +1,39 @@
...
+    # HACK: that way audit logs still land in the journal
+    AuditFilePath=/dev/null
+  '';
+
+  daemonConfFile = pkgs.writeText "usbguard-daemon-conf" daemonConf;
+
+in
+{
 
   ###### interface
```
Comment (aanderse) [inline nixos/modules/services/security/usbguard.nix:38]:
```suggestion
  imports = [
    (mkRemovedOptionModule [ "services" "usbguard" "ruleFile" ] "<explanation here>")
    (mkRemovedOptionModule [ "services" "usbgaurd" "IPCAccessControlFiles" ] "<explanation here>")
    (mkRemovedOptionModule [ "services" "usbgaurd" "auditFilePath" ] "<explanation here>")
  ];
  
  ###### interface
```

File: nixos/modules/services/security/usbguard.nix [hunk 2]
```diff
@@ -49,39 +51,27 @@ in {
...
         example = ''
           allow with-interface equals { 08:*:* }
         '';
         description = ''
-          The USBGuard daemon will load this policy rule set. Modifying it via
-          the IPC interface won't work if you use this option, since the
-          contents of this option will be written into the nix-store it will be
-          read-only.
+          The USBGuard daemon will load this to load the policy rule set.
+          Persistent modification it via IPC interface won't work, since
```
Comment (aanderse) [inline nixos/modules/services/security/usbguard.nix:62]: Typo? Missing a word?

File: nixos/modules/services/security/usbguard.nix [hunk 3]
```diff
@@ -49,39 +51,27 @@ in {
...
-        '';
-      };
-
       rules = mkOption {
         type = types.nullOr types.lines;
         default = null;
         example = ''
           allow with-interface equals { 08:*:* }
         '';
         description = ''
```
Comment (aanderse) [inline nixos/modules/services/security/usbguard.nix:60]: I find this description to be a little bit confusing, and I know what you are trying to do. Maybe we can reword this to simplify the intention of nixos managed rules which are immutable vs user managed "dynamic" rules via IPC.

**Review by aanderse — CHANGES_REQUESTED** [review]
The code looks good. Just a few things around "user education".

File: nixos/modules/services/security/usbguard.nix (hunk 1 above)
Comment (aanderse) [reply nixos/modules/services/security/usbguard.nix:38]: Hmm... further to this we will need an entry in the release notes as this is a breaking change.

**Comment by aanderse** [comment]
ping @tnias

**Comment by tnias** [comment]
@aanderse Like this?

**Review by aanderse — APPROVED** [review]
LGTM :+1: I'll defer to @flokli for final review and merge.
Thanks @tnias :tada:

**Comment by tnias** [comment]
Nice. I squashed the commits.

**Comment by flokli** [comment]
This needs another rebase unfortunately.

**Comment by tnias** [comment]
@flokli I squashed them into one. Anything else?

**Comment by flokli** [comment]
Did you rebase it onto latest master? GitHub says there's conflicts.

**Comment by tnias** [comment]
Okay, this time I actually rebased against the upstream master :laughing:

**Comment by flokli** [comment]
Thanks!
